pkg/mimirpb/split.go

```diff
@@ -0,0 +1,201 @@
+		tagVal, tagSize := binary.Uvarint(writeRequest)
+		if tagSize <= 0 {
+			return nil, io.ErrUnexpectedEOF
+		}
+
+		// totalFieldSize is updated later depending on the type and covers entire field, starting with tag.
+		totalFieldSize := tagSize
+
+		switch tagVal {
+
+		// these are wireTypeLen types, there's LENGTH (int32 varint) and then submessage with that length.
+		case timeseriesFieldTag, metadataFieldTag:
+			decodedLength, decodedLengthSize := binary.Uvarint(writeRequest[tagSize:])
+			if decodedLengthSize <= 0 {
+				return nil, io.ErrUnexpectedEOF
+			}
+			if decodedLength <= 0 || decodedLength > math.MaxInt32 {
+				return nil, fmt.Errorf("invalid decoded length: %d", decodedLength)
+			}
+			if len(writeRequest) < int(decodedLength) {
+				return nil, fmt.Errorf("length too big: %d", decodedLength)
+			}
+
+			totalFieldSize += decodedLengthSize + int(decodedLength)
+
+		// these are wireTypeVarint, value is just varint.
+		case sourceFieldTag, skipLabelNameValidationFieldTag:
+			val, valSize := binary.Uvarint(writeRequest[tagSize:])
+			if valSize <= 0 {
+				return nil, io.ErrUnexpectedEOF
+			}
+
+			if tagVal == sourceFieldTag { // Source, any int32 value is allowed.
+				if val > math.MaxInt32 {
+					return nil, fmt.Errorf("invalid value %d for tag %d", val, tagVal)
+				}
+				hasSource = true
+				source = int32(val)
+				sourceSize = valSize
+			}
+
+			if tagVal == skipLabelNameValidationFieldTag {
+				hasSkipLabelNameValidation = true
+				if val == 0 {
+					skipLabelNameValidation = false
+				} else {
+					skipLabelNameValidation = true
+				}
+			}
+
+			totalFieldSize += valSize
+
+		default:
+			// We can't handle unexpected fields.
+			return nil, fmt.Errorf("unexpected tag %d", tagVal)
+		}
+
+		// index to current subrequest, always the last one in the slice.
+		currIx := len(subrequests) - 1
+		currSize := len(subrequests[currIx])
+		// We want to extend current subrequest with this field, if possible. If not, start new subrequest with this field only.
+		if currSize+totalFieldSize <= sizeLimit {
+			subrequests[currIx] = subrequests[currIx][:currSize+totalFieldSize]
+		} else {
+			subrequests = append(subrequests, writeRequest[:totalFieldSize])
+		}
+
+		writeRequest = writeRequest[totalFieldSize:]
+	}
+
+	// We have all subrequests, now we need to post-process them in case there was source or skipLabelNameValidation field.
+	// If there were, we add them to ALL subrequests.
+	//
+	// We may end up with subrequests with multiple source or skipLabelNameValidation fields,
+	// but that's totally fine -- parsers are expected to honor last field value.
+
+	extraBytes := 0
+	if hasSource {
+		extraBytes += sourceFieldTagSize + sourceSize
+	}
+	if hasSkipLabelNameValidation {
+		extraBytes += skipLabelNameValidationFieldTagSize + boolValueVarintLength
+	}
+
+	if extraBytes > 0 {
+		for ix := range subrequests {
+			// Clone subrequest before appending bytes to it. Here we could use a buffer pool.
```

Comment:
i'm not super sure about this. It will double allocations right when we're receiving large requests in the first place.

Instead when we are marshalling the original request can we provide a larger slice: `req.MarshalToSizedBuffer()`. And then at every new subrequest we'd leave enough capacity in the slice to fit in one `source` and one `skipLabelName...`. When we're done splitting the requests we go back over the slice of subrequests and append the extra bytes if they are necessary. This way we still do the 2x copying, but at least save on the allocations.

Am I overcomplicating it?

Reply:
To put it another way - how much cheaper is to do this binary hacks+double allocation+copying vs just creating multiple `WriteRequest`s and calling `Marshall()` on each of them?

Reply:
Copied from Slack: Note that [Prometheus only ever writes field=1 or 3](https://prometheus.io/docs/concepts/remote_write_spec/#protocol). `source` is our internal field, and we only set it to non-default values from rulers (default value, API=0, is NOT serialized). Similarly, `skipLabelNameValidation` is our internal field, set by our enterprise proxies. That means that in vast majority of cases, this extra copying will never happen.